File: Java\ implementation/Book.java
// Book.java - Book model for Java implementation

/**
 * Book class represents a book in the library inventory.
 * 
 * Time Complexity:
 * - Construction: O(1)
 * - decrease/increase quantity: O(1)
 * - is_available: O(1)
 */
public class Book {
    private String bookId;      // ISBN
    private String title;
    private String author;
    private int year;
    private int quantity;
    
    /**
     * Constructor for Book
     * 
     * @param bookId Unique book identifier (ISBN)
     * @param title Book title
     * @param author Book author
     * @param year Publication year
     * @param quantity Available quantity
     */
    public Book(String bookId, String title, String author, int year, int quantity) {
        if (quantity < 0) {
            throw new IllegalArgumentException("Quantity cannot be negative");
        }
        if (year < 0 || year > 2100) {
            throw new IllegalArgumentException("Invalid publication year");
        }
        
        this.bookId = bookId;
        this.title = title;
        this.author = author;
        this.year = year;
        this.quantity = quantity;
    }
    
    // Getters
    public String getBookId() { return bookId; }
    public String getTitle() { return title; }
    public String getAuthor() { return author; }
    public int getYear() { return year; }
    public int getQuantity() { return quantity; }
    
    /**
     * Decrease book quantity.
     * 
     * Time Complexity: O(1)
     * 
     * @param amount Amount to decrease
     * @return true if successful, false if insufficient quantity
     */
    public boolean decreaseQuantity(int amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative");
        }
        if (quantity >= amount) {
            quantity -= amount;
            return true;
        }
        return false;
    }
    
    /**
     * Increase book quantity.
     * 
     * Time Complexity: O(1)
     * 
     * @param amount Amount to increase
     */
    public void increaseQuantity(int amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative");
        }
        quantity += amount;
    }
    
    /**
     * Check if book is available for borrowing.
     * 
     * Time Complexity: O(1)
     * 
     * @return true if quantity > 0
     */
    public boolean isAvailable() {
        return quantity > 0;
    }
    
    @Override
    public String toString() {
        return String.format("%s by %s (%d) - Qty: %d", 
                           title, author, year, quantity);
    }
}

File: Python implementation/models/book.py
"""
Book model for Library Management System

Represents a book in the library inventory with all necessary attributes.
"""

from dataclasses import dataclass
from typing import Optional


@dataclass
class Book:
    """
    Book data model.

    Attributes:
        book_id (str): Unique identifier (ISBN)
        title (str): Book title
        author (str): Book author
        year (int): Publication year
        quantity (int): Available quantity
    """
    book_id: str
    title: str
    author: str
    year: int
    quantity: int

    def __post_init__(self):
        """Validate data after initialization."""
        if self.quantity < 0:
            raise ValueError("Quantity cannot be negative")
        if self.year < 0 or self.year > 2100:
            raise ValueError("Invalid publication year")

    def decrease_quantity(self, amount: int = 1) -> bool:
        """
        Decrease book quantity.

        Args:
            amount: Amount to decrease (default: 1)

        Returns:
            bool: True if successful, False if insufficient quantity
        """
        if self.quantity >= amount:
            self.quantity -= amount
            return True
        return False

    def increase_quantity(self, amount: int = 1) -> None:
        """
        Increase book quantity.

        Args:
            amount: Amount to increase (default: 1)
        """
        self.quantity += amount

    def is_available(self) -> bool:
        """
        Check if book is available for borrowing.

        Returns:
            bool: True if quantity > 0
        """
        return self.quantity > 0

    def __str__(self) -> str:
        """String representation of the book."""
        return f"{self.title} by {self.author} ({self.year}) - Qty: {self.quantity}"

    def __repr__(self) -> str:
        """Detailed string representation."""
        return f"Book(id='{self.book_id}', title='{self.title}', author='{self.author}', year={self.year}, quantity={self.quantity})"